File: engines/jsc/src/promise.rs
#[cfg(feature = "native")]
use crate::ffi;

#[cfg(feature = "native")]
pub struct JSCPromise {
    inner: *mut ffi::JSCValueHandle,
}

#[cfg(feature = "native")]
impl JSCPromise {
    pub fn new(engine: &crate::ffi::JSCEnginePtr) -> Result<Self, String> {
        let ptr = engine.promise_new()?;
        Ok(Self { inner: ptr })
    }

    pub fn resolve(&self, engine: &crate::ffi::JSCEnginePtr, value: *mut ffi::JSCValueHandle) -> Result<(), String> {
        engine.promise_resolve(self.inner, value)
    }

    pub fn reject(&self, engine: &crate::ffi::JSCEnginePtr, reason: &str) -> Result<(), String> {
        engine.promise_reject(self.inner, reason)
    }

    pub fn handle(&self) -> *mut ffi::JSCValueHandle {
        self.inner
    }
}

#[cfg(feature = "native")]
impl Drop for JSCPromise {
    fn drop(&mut self) {
        if !self.inner.is_null() {
            unsafe { ffi::klyron_jsc_value_dispose(self.inner) }
        }
    }
}

File: engines/jsc/ts/promise.ts
import type { NativeJSCBindings } from "./engine.ts";

export class JSCPromise {
  constructor(
    private handle: number,
    private bindings: NativeJSCBindings,
  ) {}

  new(): number {
    return this.bindings.jscPromiseNew(this.handle);
  }

  resolve(promiseHandle: number, valueHandle: number): void {
    const r = this.bindings.jscPromiseResolve(this.handle, promiseHandle, valueHandle);
    if (!r.success) throw new Error("promiseResolve failed");
  }

  reject(promiseHandle: number, reason: string): void {
    const r = this.bindings.jscPromiseReject(this.handle, promiseHandle, reason);
    if (!r.success) throw new Error("promiseReject failed");
  }
}
